lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -935,11 +935,33 @@ static Value getCommonOperand(Op op) {
 template <typename Op>
 static bool canonicalizeIdempotentInputs(Op op, PatternRewriter &rewriter) {
   auto inputs = op.getInputs();
-  llvm::SmallSetVector<Value, 8> uniqueInputs;
 
-  for (const auto input : inputs)
-    uniqueInputs.insert(input);
+  llvm::SmallSetVector<Value, 8> alreadyCovered;
+  llvm::SmallSetVector<Value, 8> checked;
+  checked.insert(op);
 
+  llvm::SmallVector<Value, 8> worklist;
+  for (auto input : inputs) {
```

Comment:
Make sure `op` doesn't make it into the worklist.

Reply:
I think this makes sense! Having a set of checked nodes around to prevent recursion is a good pattern :smiley: Let me add a few comments.